### PR TITLE
Add metasm format of code for Super small Shell Bind TCP Random Port 

### PR DESCRIPTION
**Pull request by 3AleXM8B40**
The shellcode is going to upgrade/complement the old one https://github.com/rapid7/metasploit-framework/blob/master/modules/payloads/singles/linux/x86/shell_bind_tcp_random_port.rb which size is 57. The functionality is virtually the same but the size if SIGNIFICANTLY (44 against 57!) smaller which makes it unique in its kind.

This payload is really tiny since it uses nc by opening a random port.
Use nmap to discover the open port: 'nmap -sS -p- target'.
More details can be found here https://www.exploit-db.com/exploits/41631

Tested on Linux 4.9.0-7-amd64 #1 SMP Debian 4.9.110-1 (2018-07-05) x86_64 GNU/Linux

Please review it.
Thanks in advance!

### CODE REVIEW

File: modules/payloads/singles/linux/x86/shell_bind_tcp_super_small_random_port.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
+##
+# This module requires Metasploit: http://metasploit.com/download
```
Comment (bcoles) [inline modules/payloads/singles/linux/x86/shell_bind_tcp_super_small_random_port.rb]:
```suggestion
# This module requires Metasploit: https://metasploit.com/download
```

File: modules/payloads/singles/linux/x86/shell_bind_tcp_super_small_random_port.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  def initialize(info = {})
+    super(merge_info(info,
+      'Name'          => 'Linux Command Shell, Bind TCP Random Port Inline',
+      'Description'   => %q{
+        The tiniest (46 bytes!) bind tcp shell in its class! Listen for a connection with a random port and spawn a command shell.
+        Use nmap to discover the open port: 'nmap -sS -p- target'.
+      },
+      'Author'        => 'Aleh Boitsau <infosecurity[at]ya.ru>',
+      'License'       => BSD_LICENSE,
+      'References'    => ['URL', 'https://www.exploit-db.com/exploits/41631'],
```
Comment (bcoles) [inline modules/payloads/singles/linux/x86/shell_bind_tcp_super_small_random_port.rb]:
```suggestion
      'References'    => ['EDB', '41631'],
```

**Comment by Ekzorcist** [comment]
Hello, guys!
Please reopen https://github.com/rapid7/metasploit-framework/pull/11370
to advance with the current task!

**Comment by busterb** [comment]
Reopening #11370 really isn't required for this. Taking a look now.

**Comment by Ekzorcist** [comment]
ok! Glad to hear that)!

File: modules/payloads/singles/linux/x86/shell_bind_tcp_super_small_random_port.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+  include Msf::Payload::Single
+  include Msf::Payload::Linux
+
+  def initialize(info = {})
+    super(merge_info(info,
+      'Name'          => 'Linux Command Shell, Bind TCP Random Port Inline',
+      'Description'   => %q{
+        The tiniest (46 bytes!) bind tcp shell in its class! Listen for a connection with a random port and spawn a command shell.
+        Use nmap to discover the open port: 'nmap -sS -p- target'.
```
Comment (phra) [inline modules/payloads/singles/linux/x86/shell_bind_tcp_super_small_random_port.rb]:
```suggestion
        Use nmap to discover unfiltered closed port: 'nmap -sS -p- target'.
```

with BIND TCP, you need an available, not filtered, closed port, not an open one.

**Review by phra — CHANGES_REQUESTED** [review]

File: modules/payloads/singles/linux/x86/shell_bind_tcp_super_small_random_port.rb (hunk 3 above)
Comment (Ekzorcist) [reply modules/payloads/singles/linux/x86/shell_bind_tcp_super_small_random_port.rb]: phra, yes, we need not filtered and not used (closed) port when it comes to bind tcp.

But, once shellcode triggers - it opens a port, so we need to scan a target to find that opened port in order to connect to it.

Comment (phra) [reply modules/payloads/singles/linux/x86/shell_bind_tcp_super_small_random_port.rb]: got it. :+1:

**Comment by phra** [comment]
> This payload is really tiny since it uses nc by opening a random port.

the command passed to the `execve()` is `nc -le /bin/sh` so it requires `nc.traditional` installed on the target machine making the shellcode dependent on the machine configuration.

**Comment by Ekzorcist** [comment]
> > This payload is really tiny since it uses nc by opening a random port.
> 
> the command passed to the `execve()` is `nc -le /bin/sh` so it requires `nc.traditional` installed on the target machine making the shellcode dependent on the machine configuration.

yes, I am aware of it. And this is a deliberate approach to make it the tiniest one!
Regardless of the used technique to open a port your attempt to do this can also be limited with other factors such as iptables INPUT configuration, SELINUX etc..
Nothing is universal for every case.

**Comment by Ekzorcist** [comment]
Hello, @busterb !
Is there any news on reviewing and merging the shellcode?
Appreciate your help on this a lot!!!

**Comment by busterb** [comment]
Hi @Ekzorcist yeah I started reviewing this, noticed that our payload size checker automation was actually broken!, and didn't finish. Plan to get back on this soon to fix that as well.

**Comment by Ekzorcist** [comment]
Thank you!!

**Comment by Ekzorcist** [comment]
Hello, Brent!
How is it going with payload size checker automation?
May be I could help in fixing it ?

**Comment by busterb** [comment]
All fixed, there was a bug in msfvenom's size checker that got merged in #11821. I have some changes I'm going to include with the merge here too.

**Comment by busterb** [comment]
I added 9ae01c98c9 which merged this payload into the existing bind payload. This makes it automatically invoked as needed rather than having to evaluate a different payload entirely (this pattern is in many other payloads as well.) If you pass `-s 56` or smaller to msfvenom, it will choose the tinier version, e.g. `./msfvenom -p linux/x86/shell_bind_tcp_random_port -s 57`

Note also that I had to rename `generate_bind_tcp_shell` to `generate`, since the original payload in this PR never generated anything unless it overrode the base class method for the payload class. Keep that in mind for any future efforts.

Thanks!

**Comment by acammack-r7** [comment]
# Release Notes

Adds a small version of `linux/x86/shell_bind_tcp_random_port` that uses the `nc` command on the target to reduce the amount of shellcode needed. The new payload will automatically be used when the old one is too large.

**Comment by Ekzorcist** [comment]
Super cool!

**Pull request merged**